src/string_object.cpp

```diff
@@ -1959,6 +1959,62 @@ Value StringObject::chop_in_place(Env *env) {
     return this;
 }
 
+Value StringObject::partition(Env *env, Value val) {
+    auto ary = new ArrayObject;
+
+    if (val->is_regexp()) {
+        auto match_result = val->as_regexp()->match(env, this);
+    
+        ssize_t start_byte_index;
+        ssize_t end_byte_index;
+
+        if (match_result->is_nil()){
+            auto empty = new StringObject { "" };
+            return new ArrayObject { this, empty, empty };
+        } else {
+            start_byte_index = match_result->as_match_data()->index(0);
+            end_byte_index = match_result->as_match_data()->ending(0);
+            ary->push(new StringObject(m_string.substring(0, start_byte_index)));
+        }
+
+        ary->push(new StringObject(m_string.substring(start_byte_index, end_byte_index - start_byte_index)));
+
+        auto last_char = start_byte_index + (end_byte_index - start_byte_index);
+        auto len = m_string.length() - 1;
```

Comment:
I think the naming here is confusing. This isn't `len` (length) -- it's actually the last character index. I'd prefer we use more descriptive names that match what they are, even if it means the names are longer, e.g. `last_char_index` would be fine!